src/app/shared/modules/context-menu/context-menu.service.ts

```diff
@@ -0,0 +1,130 @@
+    textSelect: new Subscription(),
+    longPress: new Subscription(),
+  };
+  /** Subscribeable list of all registered action handlers */
+  private actionHandlers$: {
+    [context in IContextMenuType]: BehaviorSubject<IContextMenuActionHashmap>;
+  } = {
+    rightClick: new BehaviorSubject({}),
+    textSelect: new BehaviorSubject({}),
+    longPress: new BehaviorSubject({}),
+  };
+
+  /** Utility variable to help detect start/end of text selection events */
+  private isSelectingText = false;
+
+  constructor(private popoverController: PopoverController) {
+    this.addEventListeners();
+  }
+
+  /** Add an item to appear in the context menu */
+  public registerContextMenuAction(context: IContextMenuType, action: IContextMenuAction) {
+    const actions = this.actionHandlers$[context].value;
+    actions[action.id] = action;
+    this.actionHandlers$[context].next(actions);
+  }
+
+  /** Remove an item from the context menu */
+  public removeContextMenuAction(context: IContextMenuType, id: string) {
+    const actions = this.actionHandlers$[context].value;
+    if (actions.hasOwnProperty(id)) {
+      delete actions[id];
+    }
+    this.actionHandlers$[context].next(actions);
+  }
+
+  private async showContextMenu(context: IContextMenuType, ev: Event, data: any = {}) {
+    // if no registered actions return (should be unsubscribed but does not always seem to manage correctly)
+    // (likely requires better register/remove in batch instead of sequentially)
+    const actions = Object.values(this.actionHandlers$[context].value);
+    if (actions.length === 0) {
+      return;
+    }
+    // if already open close before reopening
+    const existingPopover = await this.popoverController.getTop();
+    if (existingPopover) {
+      await this.popoverController.dismiss();
+    }
+    // pass registered action handlers and event as input props to display in the ContextMenuComponent
+    const componentProps = { actions, event: ev, data };
+    const popover = await this.popoverController.create({
+      component: ContextMenuComponent,
+      cssClass: "context-menu",
+      event: ev,
+      reference: "event", // create menu from where event triggered
+      translucent: true,
+      componentProps,
+      dismissOnSelect: true,
+    });
+    await popover.present();
+  }
+
+  /**
+   * Listen to right click (contextmenu) and text select (selectionchange)
+   * events to trigger event handlers
+   **/
+  private addEventListeners() {
+    const eventListeners: {
+      [context in IContextMenuType]: Observable<{ ev: PointerEvent; data?: any }>;
+    } = {
+      // Long-press (TODO)
+      longPress: fromEvent(document, "TODO").pipe(map((ev: PointerEvent) => ({ ev }))),
+      // Right-click
+      rightClick: fromEvent(document, "contextmenu").pipe(map((ev: PointerEvent) => ({ ev }))),
+      // Text select
+      // Events continously triggered so include additional mapping to detect start/end of selection
```

Comment:
text-select is an interesting/slightly more challenging situation as events are constantly triggered every time any character is changed (instead of waiting to the end of selection). This is handled using various utility methods that swap from listening to start of text selection to waiting for mouse-up event, and then retrieving selected text